File: .sling/libs/web5/eqela/jk-net/0.38.0/jk.socket.async.js
let JkSocketAsyncTCPClientForIOManagerMyBuffer = function() {
	this.data = null;
	this.size = 0;
	this.callback = null;
};

JkSocketAsyncTCPClientForIOManagerMyBuffer.prototype._t = { "JkSocketAsyncTCPClientForIOManagerMyBuffer" : true };
JkSocketAsyncTCPClientForIOManagerMyBuffer.prototype._tobj = JkSocketAsyncTCPClientForIOManagerMyBuffer;

JkSocketAsyncTCPClientForIOManagerMyBuffer.NEW = function() {
	var v = new JkSocketAsyncTCPClientForIOManagerMyBuffer;
	return v.CTOR_JkSocketAsyncTCPClientForIOManagerMyBuffer();
};

JkSocketAsyncTCPClientForIOManagerMyBuffer.prototype.CTOR_JkSocketAsyncTCPClientForIOManagerMyBuffer = function() {
	this.callback = null;
	this.size = 0;
	this.data = null;
	return this;
};

JkSocketAsyncTCPClientForIOManagerMyBuffer.prototype.getData = function() {
	return this.data;
};

JkSocketAsyncTCPClientForIOManagerMyBuffer.prototype.setData = function(v) {
	this.data = v;
	return this;
};

JkSocketAsyncTCPClientForIOManagerMyBuffer.prototype.getSize = function() {
	return this.size;
};

JkSocketAsyncTCPClientForIOManagerMyBuffer.prototype.setSize = function(v) {
	this.size = v;
	return this;
};

JkSocketAsyncTCPClientForIOManagerMyBuffer.prototype.getCallback = function() {
	return this.callback;
};

JkSocketAsyncTCPClientForIOManagerMyBuffer.prototype.setCallback = function(v) {
	this.callback = v;
	return this;
};

JkSocketAsyncTCPClientForIOManagerMyBuffer.IS_INSTANCE = function(o) {
	return o != null && o._t != null && o._t["JkSocketAsyncTCPClientForIOManagerMyBuffer"] === true;
};

let JkSocketAsyncTCPClient = function() {};

JkSocketAsyncTCPClient.prototype._t = { "JkSocketAsyncTCPClient" : true };
JkSocketAsyncTCPClient.prototype._tobj = JkSocketAsyncTCPClient;

JkSocketAsyncTCPClient.NEW = function() {
	var v = new JkSocketAsyncTCPClient;
	return v.CTOR_JkSocketAsyncTCPClient();
};

JkSocketAsyncTCPClient.prototype.CTOR_JkSocketAsyncTCPClient = function() {
	return this;
};

JkSocketAsyncTCPClient.instance = function() {
	return JkSocketAsyncTCPClientForIOManager.forCurrentIOManager();
};

JkSocketAsyncTCPClient.prototype.connectSSL = function(address, port, callback) {};

JkSocketAsyncTCPClient.prototype.connect = function(address, port, callback) {};

JkSocketAsyncTCPClient.prototype.receive = function(callback) {};

JkSocketAsyncTCPClient.prototype.send = function(data, size, callback) {};

JkSocketAsyncTCPClient.prototype.disconnect = function(callback) {};

JkSocketAsyncTCPClient.IS_INSTANCE = function(o) {
	return o != null && o._t != null && o._t["JkSocketAsyncTCPClient"] === true;
};

let JkSocketAsyncTCPClientForIOManager = function() {
	JkSocketAsyncTCPClient.call(this);
	this.tcpsocket = null;
	this.sslsocket = null;
	this.rcvbuffer = null;
	this.iomanager = null;
	this.toWrite = JkLangQueue.NEW();
	this.writeScheduled = false;
};

JkSocketAsyncTCPClientForIOManager.prototype = (function(o) { function F() {}; F.prototype = o; return(new F()); }) (JkSocketAsyncTCPClient.prototype);
JkSocketAsyncTCPClientForIOManager.prototype.constructor = JkSocketAsyncTCPClientForIOManager;
JkSocketAsyncTCPClientForIOManager.prototype._t = {
	"JkSocketAsyncTCPClientForIOManager" : true,
	"JkSocketAsyncTCPClient" : true
};
JkSocketAsyncTCPClientForIOManager.prototype._tobj = JkSocketAsyncTCPClientForIOManager;

JkSocketAsyncTCPClientForIOManager.NEW = function() {
	var v = new JkSocketAsyncTCPClientForIOManager;
	return v.CTOR_JkSocketAsyncTCPClientForIOManager();
};

JkSocketAsyncTCPClientForIOManager.prototype.CTOR_JkSocketAsyncTCPClientForIOManager = function() {
	if(JkSocketAsyncTCPClient.prototype.CTOR_JkSocketAsyncTCPClient.call(this) == null) {
		return null;
	}
	this.writeScheduled = false;
	this.toWrite = JkLangQueue.NEW();
	this.iomanager = null;
	this.rcvbuffer = null;
	this.sslsocket = null;
	this.tcpsocket = null;
	this.rcvbuffer = new ArrayBuffer(64 * 1024);
	return this;
};

JkSocketAsyncTCPClientForIOManager.forCurrentIOManager = function() {
	var mgr = JkSocketIoIOManager.getCurrent();
	if(!(mgr != null)) {
		return null;
	}
	var v = JkSocketAsyncTCPClientForIOManager.NEW();
	v.setIomanager(mgr);
	return v;
};

JkSocketAsyncTCPClientForIOManager.forIOManager = function(mgr) {
	var v = JkSocketAsyncTCPClientForIOManager.NEW();
	v.setIomanager(mgr);
	return v;
};

JkSocketAsyncTCPClientForIOManager.prototype.getSocket = function() {
	if(this.sslsocket != null) {
		return this.sslsocket;
	}
	return this.tcpsocket;
};

JkSocketAsyncTCPClientForIOManager.prototype.doConnectSocket = function(address, port) {
	return JkSocketTCPSocket.createAndConnect(address, port);
};

JkSocketAsyncTCPClientForIOManager.prototype.connectSSL = function(address, port, callback) {
	if(!((this.tcpsocket = this.doConnectSocket(address, port)) != null)) {
		if(callback != null) {
			callback((JkLangError.forCode("failedToConnect", null)));
		}
		return;
	}
	if(!((this.sslsocket = JkSocketSslSSLSocket.forClient(this.tcpsocket, address, null, false, null)) != null)) {
		if(callback != null) {
			callback((JkLangError.forCode("sslHandshakeFailed", null)));
		}
		return;
	}
	if(callback != null) {
		callback(null);
	}
};

JkSocketAsyncTCPClientForIOManager.prototype.connect = function(address, port, callback) {
	if(!((this.tcpsocket = this.doConnectSocket(address, port)) != null)) {
		if(callback != null) {
			callback((JkLangError.forCode("failedToConnect", null)));
		}
		return;
	}
	this.sslsocket = null;
	if(callback != null) {
		callback(null);
	}
};

JkSocketAsyncTCPClientForIOManager.prototype.receive = function(callback) {
	if(!(this.tcpsocket != null)) {
		if(callback != null) {
			callback(null, 0);
		}
		return;
	}
	var ee = this.iomanager.add(this.tcpsocket);
	var cc = callback;
	ee.setReadListener((function() {
		ee.setReadListener(null);
		var ss = this.getSocket();
		var bb = this.rcvbuffer;
		var v = 0;
		if(ss != null) {
			v = ss.read(bb);
		}
		if(cc != null) {
			cc(bb, v);
		}
	}.bind(this)));
};

JkSocketAsyncTCPClientForIOManager.prototype.send = function(data, size, callback) {
	if(!(this.tcpsocket != null)) {
		if(callback != null) {
			callback((JkLangError.forCode("notConnected", null)));
		}
		return;
	}
	var mb = JkSocketAsyncTCPClientForIOManagerMyBuffer.NEW();
	mb.setData(data);
	mb.setSize(size);
	mb.setCallback(callback);
	this.toWrite.push(mb);
	this.scheduleWrite();
};

JkSocketAsyncTCPClientForIOManager.prototype.scheduleWrite = function() {
	if(this.writeScheduled) {
		return;
	}
	if(this.toWrite.getSize() < 1) {
		return;
	}
	var ee = this.iomanager.add(this.tcpsocket);
	ee.setWriteListener((function() {
		ee.setWriteListener(null);
		this.writeScheduled = false;
		var mb = this.toWrite.pop();
		if(mb != null) {
			var data = mb.getData();
			var size = mb.getSize();
			var callback = mb.getCallback();
			var ss = this.getSocket();
			var n = 0;
			if(ss != null) {
				n = ss.write(data, size);
			}
			if(callback != null) {
				if(n < size) {
					callback((JkLangError.forCode("failedToWrite", null)));
				}
				else {
					callback(null);
				}
			}
		}
		this.scheduleWrite();
	}.bind(this)));
	this.writeScheduled = true;
};

JkSocketAsyncTCPClientForIOManager.prototype.disconnect = function(callback) {
	this.toWrite.clear();
	this.writeScheduled = false;
	if(this.sslsocket != null) {
		this.sslsocket.close();
		this.sslsocket = null;
	}
	if(this.tcpsocket != null) {
		this.tcpsocket.close();
		this.tcpsocket = null;
	}
	if(callback != null) {
		callback(null);
	}
};

JkSocketAsyncTCPClientForIOManager.prototype.getIomanager = function() {
	return this.iomanager;
};

JkSocketAsyncTCPClientForIOManager.prototype.setIomanager = function(v) {
	this.iomanager = v;
	return this;
};

JkSocketAsyncTCPClientForIOManager.IS_INSTANCE = function(o) {
	return o != null && o._t != null && o._t["JkSocketAsyncTCPClientForIOManager"] === true;
};

File: .sling/libs/sushi/eqela/jk-net/0.27.0/jk.socket.async.lua
jk = jk or {}

jk.socket = jk.socket or {}

jk.socket.async = jk.socket.async or {}

jk.socket.async.TCPClient = {}
jk.socket.async.TCPClient.__index = jk.socket.async.TCPClient
_vm:set_metatable(jk.socket.async.TCPClient, {})

function jk.socket.async.TCPClient._create()
	local v = _vm:set_metatable({}, jk.socket.async.TCPClient)
	return v
end

function jk.socket.async.TCPClient:_init()
	self._isClassInstance = true
	self._qualifiedClassName = self._qualifiedClassName or 'jk.socket.async.TCPClient'
	self['_isType.jk.socket.async.TCPClient'] = true
end

function jk.socket.async.TCPClient:_construct0()
	jk.socket.async.TCPClient._init(self)
	return self
end

function jk.socket.async.TCPClient:instance()
	do return _g.jk.socket.async.TCPClientForIOManager:forCurrentIOManager() end
end

function jk.socket.async.TCPClient:connectSSL(address, port, callback)
end

function jk.socket.async.TCPClient:connect(address, port, callback)
end

function jk.socket.async.TCPClient:receive(callback)
end

function jk.socket.async.TCPClient:send(data, size, callback)
end

function jk.socket.async.TCPClient:disconnect(callback)
end

jk.socket.async.TCPClientForIOManager = _g.jk.socket.async.TCPClient._create()
jk.socket.async.TCPClientForIOManager.__index = jk.socket.async.TCPClientForIOManager
_vm:set_metatable(jk.socket.async.TCPClientForIOManager, {
	__index = _g.jk.socket.async.TCPClient
})

function jk.socket.async.TCPClientForIOManager._create()
	local v = _vm:set_metatable({}, jk.socket.async.TCPClientForIOManager)
	return v
end

function jk.socket.async.TCPClientForIOManager:_init()
	self._isClassInstance = true
	self._qualifiedClassName = self._qualifiedClassName or 'jk.socket.async.TCPClientForIOManager'
	self['_isType.jk.socket.async.TCPClientForIOManager'] = true
	self.tcpsocket = nil
	self.sslsocket = nil
	self.rcvbuffer = nil
	self.iomanager = nil
	self.toWrite = _g.jk.lang.Queue._construct0(_g.jk.lang.Queue._create())
	self.writeScheduled = false
end

function jk.socket.async.TCPClientForIOManager:forCurrentIOManager()
	local mgr = _g.jk.socket.io.IOManager:getCurrent()
	if not (mgr ~= nil) then
		do return nil end
	end
	do
		local v = _g.jk.socket.async.TCPClientForIOManager._construct0(_g.jk.socket.async.TCPClientForIOManager._create())
		do v:setIomanager(mgr) end
		do return v end
	end
end

function jk.socket.async.TCPClientForIOManager:forIOManager(mgr)
	local v = _g.jk.socket.async.TCPClientForIOManager._construct0(_g.jk.socket.async.TCPClientForIOManager._create())
	do v:setIomanager(mgr) end
	do return v end
end

function jk.socket.async.TCPClientForIOManager:_construct0()
	jk.socket.async.TCPClientForIOManager._init(self)
	do _g.jk.socket.async.TCPClient._construct0(self) end
	self.rcvbuffer = _util:allocate_buffer(64 * 1024)
	return self
end

function jk.socket.async.TCPClientForIOManager:getSocket()
	if self.sslsocket ~= nil then
		do return self.sslsocket end
	end
	do return self.tcpsocket end
end

function jk.socket.async.TCPClientForIOManager:doConnectSocket(address, port)
	do return _g.jk.socket.TCPSocket:createAndConnect(address, port) end
end

function jk.socket.async.TCPClientForIOManager:connectSSL(address, port, callback)
	if not ((function()
		self.tcpsocket = self:doConnectSocket(address, port)
		do return self.tcpsocket end
	end)() ~= nil) then
		if callback ~= nil then
			do callback(_g.jk.lang.Error:forCode("failedToConnect", nil)) end
		end
		do return end
	end
	if not ((function()
		self.sslsocket = _g.jk.socket.ssl.SSLSocket:forClient(self.tcpsocket, address, nil, false, nil)
		do return self.sslsocket end
	end)() ~= nil) then
		if callback ~= nil then
			do callback(_g.jk.lang.Error:forCode("sslHandshakeFailed", nil)) end
		end
		do return end
	end
	if callback ~= nil then
		do callback(nil) end
	end
end

function jk.socket.async.TCPClientForIOManager:connect(address, port, callback)
	if not ((function()
		self.tcpsocket = self:doConnectSocket(address, port)
		do return self.tcpsocket end
	end)() ~= nil) then
		if callback ~= nil then
			do callback(_g.jk.lang.Error:forCode("failedToConnect", nil)) end
		end
		do return end
	end
	self.sslsocket = nil
	if callback ~= nil then
		do callback(nil) end
	end
end

function jk.socket.async.TCPClientForIOManager:receive(callback)
	if not (self.tcpsocket ~= nil) then
		if callback ~= nil then
			do callback(nil, 0) end
		end
		do return end
	end
	do
		local ee = self.iomanager:add(self.tcpsocket)
		local cc = callback
		do ee:setReadListener(function()
			do ee:setReadListener(nil) end
			do
				local ss = self:getSocket()
				local bb = self.rcvbuffer
				local v = 0
				if ss ~= nil then
					v = ss:read(bb)
				end
				if cc ~= nil then
					do cc(bb, v) end
				end
			end
		end) end
	end
end

jk.socket.async.TCPClientForIOManager.MyBuffer = {}
jk.socket.async.TCPClientForIOManager.MyBuffer.__index = jk.socket.async.TCPClientForIOManager.MyBuffer
_vm:set_metatable(jk.socket.async.TCPClientForIOManager.MyBuffer, {})

function jk.socket.async.TCPClientForIOManager.MyBuffer._create()
	local v = _vm:set_metatable({}, jk.socket.async.TCPClientForIOManager.MyBuffer)
	return v
end

function jk.socket.async.TCPClientForIOManager.MyBuffer:_init()
	self._isClassInstance = true
	self._qualifiedClassName = self._qualifiedClassName or 'jk.socket.async.TCPClientForIOManager.MyBuffer'
	self['_isType.jk.socket.async.TCPClientForIOManager.MyBuffer'] = true
	self.data = nil
	self.size = 0
	self.callback = nil
end

function jk.socket.async.TCPClientForIOManager.MyBuffer:_construct0()
	jk.socket.async.TCPClientForIOManager.MyBuffer._init(self)
	return self
end

function jk.socket.async.TCPClientForIOManager.MyBuffer:getData()
	do return self.data end
end

function jk.socket.async.TCPClientForIOManager.MyBuffer:setData(v)
	self.data = v
	do return self end
end

function jk.socket.async.TCPClientForIOManager.MyBuffer:getSize()
	do return self.size end
end

function jk.socket.async.TCPClientForIOManager.MyBuffer:setSize(v)
	self.size = v
	do return self end
end

function jk.socket.async.TCPClientForIOManager.MyBuffer:getCallback()
	do return self.callback end
end

function jk.socket.async.TCPClientForIOManager.MyBuffer:setCallback(v)
	self.callback = v
	do return self end
end

function jk.socket.async.TCPClientForIOManager:send(data, size, callback)
	if not (self.tcpsocket ~= nil) then
		if callback ~= nil then
			do callback(_g.jk.lang.Error:forCode("notConnected", nil)) end
		end
		do return end
	end
	do
		local mb = _g.jk.socket.async.TCPClientForIOManager.MyBuffer._construct0(_g.jk.socket.async.TCPClientForIOManager.MyBuffer._create())
		do mb:setData(data) end
		do mb:setSize(size) end
		do mb:setCallback(callback) end
		do self.toWrite:push(mb) end
		do self:scheduleWrite() end
	end
end

function jk.socket.async.TCPClientForIOManager:scheduleWrite()
	if self.writeScheduled then
		do return end
	end
	if self.toWrite:getSize() < 1 then
		do return end
	end
	do
		local ee = self.iomanager:add(self.tcpsocket)
		do ee:setWriteListener(function()
			do ee:setWriteListener(nil) end
			self.writeScheduled = false
			do
				local mb = self.toWrite:pop()
				if mb ~= nil then
					local data = mb:getData()
					local size = mb:getSize()
					local callback = mb:getCallback()
					local ss = self:getSocket()
					local n = 0
					if ss ~= nil then
						n = ss:write(data, size)
					end
					if callback ~= nil then
						if n < size then
							do callback(_g.jk.lang.Error:forCode("failedToWrite", nil)) end
						else
							do callback(nil) end
						end
					end
				end
				do self:scheduleWrite() end
			end
		end) end
		self.writeScheduled = true
	end
end

function jk.socket.async.TCPClientForIOManager:disconnect(callback)
	do self.toWrite:clear() end
	self.writeScheduled = false
	if self.sslsocket ~= nil then
		do self.sslsocket:close() end
		self.sslsocket = nil
	end
	if self.tcpsocket ~= nil then
		do self.tcpsocket:close() end
		self.tcpsocket = nil
	end
	if callback ~= nil then
		do callback(nil) end
	end
end

function jk.socket.async.TCPClientForIOManager:getIomanager()
	do return self.iomanager end
end

function jk.socket.async.TCPClientForIOManager:setIomanager(v)
	self.iomanager = v
	do return self end
end
